src/query/api/v1/httpd/handler.go

```diff
@@ -318,19 +310,78 @@ func (h *Handler) RegisterRoutes() error {
 					Tagged(remoteSource).
 					Tagged(experimentalAPIGroup),
 			)
-			h.router.HandleFunc(annotated.WriteURL,
-				wrapped(experimentalAnnotatedWriteHandler).ServeHTTP,
-			).Methods(annotated.WriteHTTPMethod)
+			wrappedRouteFn(annotated.WriteURL, experimentalAnnotatedWriteHandler,
+				annotated.WriteHTTPMethod)
 		}
 	}
 
 	h.registerHealthEndpoints()
 	h.registerProfileEndpoints()
 	h.registerRoutesEndpoint()
 
+	// Register custom endpoints.
+	for _, custom := range h.customHandlers {
+		for _, method := range custom.Methods() {
+			routeName := routeName(custom.Route(), method)
+			route := h.router.Get(routeName)
+			var prevHandler http.Handler
+			if route != nil {
+				prevHandler = route.GetHandler()
+			}
+			customHandler, err := custom.Handler(nativeSourceOpts, prevHandler)
+			if err != nil {
+				return fmt.Errorf("failed to register custom handler with path %s: %w",
+					routeName, err)
+			}
+
+			if route == nil {
+				wrappedRouteFn(custom.Route(), customHandler, method)
+			} else {
+				route.Handler(wrapped(customHandler))
+			}
+		}
+	}
+
 	return nil
 }
 
+func (h *Handler) addRoute(
+	router *mux.Router,
+	path string,
+	handler http.Handler,
+	methods []string,
+) {
+	h.addRouteHandlerFn(router, path, handler.ServeHTTP, methods...)
+}
+
+func (h *Handler) addRouteHandlerFn(
+	router *mux.Router,
+	path string,
+	handlerFn http.HandlerFunc,
+	methods ...string,
+) {
+	for _, method := range methods {
```

Comment:
I think the previous way of deduping methods then passing them all into the `router.Methods()` was clearer

Reply:
But the previous way had some issues as @vpranckaitis mentioned in his [comment](https://github.com/m3db/m3/pull/2872#discussion_r521951297).
The behaviour should be exactly the same as before, we just improved our ability to correctly detect duplicates. Previous approach wouldn't detect duplicate if e.g. route was added with GET and POST methods and then custom handler tried to override the same route with only GET method.